python/pyspark/sql/types.py

```diff
@@ -507,20 +503,52 @@ def add(self, field, data_type=None, nullable=True, metadata=None):
             else:
                 data_type_f = data_type
             self.fields.append(StructField(field, data_type_f, nullable, metadata))
-            self.names.append(field)
-        self._needSerializeAnyField = any(f.needConversion() for f in self.fields)
         return self
 
+    @property
+    def names(self):
+        """Return a list of the field names"""
+        return [field.name for field in self]
+
+    @property
+    def _needSerializeAnyField(self):
+        """Determine if any field needs conversion"""
+        return any(field.needConversion() for field in self)
+
+    def __iter__(self):
+        """Iterate the fields"""
+        return iter(self.fields)
+
+    def __len__(self):
+        """Return the number of fields."""
+        return len(self.fields)
+
+    def __getitem__(self, key):
+        """Access fields by name or slice."""
+        if isinstance(key, str):
+            _dict_fields = {field.name: field for field in self}
```

Comment:
So looking some more at the Scala implementation, this is done in such a way that the order of the fields returned is in the order of the fields in the schema  - it would be good to preserve this behaviour in the Python API for consistencies sake (and this implementation might I'm not super sure on Python slicing behaviour with dictionaries) but it would probably be good to add a test for this behaviour as well. What do you think?